luatest/assertions.lua

```diff
@@ -85,12 +86,94 @@ local function error_msg_equality(actual, expected, deep_analysis)
 end
 M.private.error_msg_equality = error_msg_equality
 
+--
+-- return args as table with 'n' set to args number
+--
+local function table_pack(...)
+    return {n = select('#', ...), ...}
+end
+
+--
+-- The wrapper is used when trace check is required. See pcall_check_trace.
+--
+-- Without wrapper the trace will point to the pcall implementation. So trace
+-- check is not strict enough (the trace can point to any pcall in below in
+-- call trace).
+--
+local trace_line = debug.getinfo(1, 'l').currentline + 2
+local function wrapped_call(fn, ...)
+    local res = table_pack(fn(...))
+    -- With `return fn(...)` wrapper does not work due to tail call
+    -- optimization.
+    return unpack(res, 1, res.n)
+end
+
+-- Expected trace for trace check. See pcall_check_trace.
+local wrapped_trace = {
+    file = debug.getinfo(1, 'S').short_src,
+    line = trace_line,
+}
+
+-- Used in tests to force check for given module.
+M.private.check_trace_module = nil
+
+--
+-- Return true if error trace check is required for function. Basically it is
+-- just a wrapper around Tarantool's utils.proper_trace_required. Additionally
+-- old Tarantool versions where this function is not present are handled.
+--
+local function trace_check_is_required(fn)
+    local src = debug.getinfo(fn, 'S').short_src
+    if M.private.check_trace_module == src then
```

Comment:
Unfortunately it will not work. The issue is we use `assert_error()` to test failure on trace check. For example for below code we fail because the error raised by `t.assert_error_msg_equals` on trace check failure is not a box error. It will be checked by `assert_failure_contains()` which is a wrapper on `assert_error()`.

```lua
    assert_failure_contains('Unexpected error trace, expected:',
                            t.assert_error_msg_equals, 'bar', f_check_trace,
                            'bar', 1)

```